### PR TITLE
ci(release): Refactor release script with reusable workflow

### PR DESCRIPTION
**Pull request by Boshen**
## Summary

This PR removes code duplication around build tasks, which comes in handy when we separate out all the tasks for canary, nightly and stable releases.

This PR also does the following:
* change engines.pnpm to >=7.25.0 is it doesn't fail the build in docker
* builds not being cached across runs - we need to cache `node_binding/target`
* adds nightly run

Successful Runs:
* PR https://github.com/web-infra-dev/rspack/actions/runs/4685119642/jobs/8302069340
* Main https://github.com/web-infra-dev/rspack/actions/runs/4685119641/jobs/8302140079
* Nightly https://github.com/web-infra-dev/rspack/actions/runs/4685119644/jobs/8302231143

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [ ] New feature / Improvement
- [x] Refactoring
- [ ] Breaking change

## Checklist

- [ ] I have added changeset via `pnpm run changeset`.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -486,18 +349,18 @@ jobs:
...
+          # version: ${{ github.event.inputs.version }}
+          # type: "release"
+          # branch: ${{ github.ref_name }}
+          # tools: "changeset"
+        # env:
+          # GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          # NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+          # REPOSITORY: ${{ github.repository }}
+          # REF: ${{ github.ref }}
+          # ONLY_RELEASE_TAG: true
```
Comment (Boshen) [inline .github/workflows/release.yml:363]: This is broken on main

Comment (Boshen) [reply .github/workflows/release.yml:363]: ping @IWANABETHATGUY

Comment (IWANABETHATGUY) [reply .github/workflows/release.yml:363]: could you share the broken CI log link?

Comment (Boshen) [reply .github/workflows/release.yml:304]: This is tacked onto a pnpm run, maybe you meant `uses: modern-js-dev/actions@v2`.

You can take a look at the current Release runs, they are all failing.

Comment (IWANABETHATGUY) [reply .github/workflows/release.yml:304]: you are right

**Comment by hyf0** [comment]
I comment on the release-pre-check in https://github.com/web-infra-dev/rspack/pull/2662 due to blocking the current release. You might need to add it back in this PR.

File: .github/workflows/test-pr.yml [hunk 2]
```diff
@@ -1,21 +1,24 @@
...
+    # paths:
+      # - "crates/**"
+      # - "packages/**"
+      # - "Cargo.toml"
+      # - "Cargo.lock"
+      # - "rust-toolchain.toml"
+      # - "package.json"
+      # - "pnpm-lock.yaml"
+      # - "pnpm-workspace.yaml"
+    # branches:
```
Comment (h-a-n-a) [inline .github/workflows/test-pr.yml:17]: Disabling these seems not expected?

Comment (Boshen) [reply .github/workflows/test-pr.yml:17]: I'll uncomment it once everything else is good. This is used for testing.

Comment (h-a-n-a) [reply .github/workflows/test-pr.yml:17]: LGTM

**Review by hardfist — APPROVED** [review]

File: .github/workflows/reusable-build.yml [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+          PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD: true
+          PUPPETEER_SKIP_DOWNLOAD: true
+        shell: bash
+        run: |
+          set -e
+          corepack enable
+          pnpm install
+          pnpm run build:js
+          pnpm run test:unit
+          # These two breaks on many targets, diabled for now
```
Comment (h-a-n-a) [inline .github/workflows/reusable-build.yml:284]: I think at least you can keep these on this platform?

**Pull request merged**